scale/batch/models.py

```diff
@@ -0,0 +1,299 @@
+        """
+
+        # Fetch the requested batch for processing
+        batch = Batch.objects.get(pk=batch_id)
+        if batch.status == 'CREATED':
+            raise BatchError('Batch already completed: %i', batch_id)
+        batch_definition = batch.get_batch_definition()
+
+        # Fetch all the recipes of the requested type that are not already superseded
+        old_recipes = Recipe.objects.filter(recipe_type=batch.recipe_type, is_superseded=False)
+
+        # Exclude recipes that have not actually changed unless requested
+        if not (batch_definition.job_names or batch_definition.all_jobs):
+            old_recipes = old_recipes.filter(recipe_type__revision_num__gt=F('recipe_type_rev__revision_num'))
+
+        # Optionally filter by date range
+        if batch_definition.started:
+            old_recipes = old_recipes.filter(created__gte=batch_definition.started)
+        if batch_definition.ended:
+            old_recipes = old_recipes.filter(created__lte=batch_definition.ended)
+
+        total = old_recipes.count()
+        logger.info('Scheduling batch recipes: %i', total)
+        batch.total_count = total
+        batch.save()
+
+        # Schedule all the new recipes/jobs and create corresponding batch models
+        for old_recipe in old_recipes.iterator():
+            try:
+                self._process_recipe(batch, old_recipe)
+            except:
+                logger.exception('Unable to supersede batch recipe: %i', old_recipe.id)
+                batch.failed_count += 1
+                batch.save()
+        logger.info('Created: %i, Failed: %i', batch.created_count, batch.failed_count)
+
+        # Update the final batch state
+        # Recompute the total to catch models that may have matched after the count query
+        batch.status = 'CREATED'
+        batch.total_count = batch.created_count + batch.failed_count
+        batch.save()
+
+    @transaction.atomic
+    def _process_recipe(self, batch, old_recipe):
+        """Creates all the batch-specific models to track the new jobs that were queued.
+
+        Each batch recipe and its batch jobs are created in an atomic transaction to support resuming the batch command
+        when it is interrupted prematurely.
+
+        :param batch: The batch that defines the recipes to schedule
+        :type batch: :class:`batch.models.Batch`
+        :param old_recipe: The old recipe that was superseded
+        :type old_recipe: :class:`recipe.models.Recipe`
+        """
+
+        # Check whether the batch recipe already exists
+        if BatchRecipe.objects.filter(batch=batch, superseded_recipe=old_recipe).exists():
+            return
+
+        # Create the new recipe and its associated jobs
+        batch_definition = batch.get_batch_definition()
+        handler = Recipe.objects.reprocess_recipe(old_recipe.id, batch_definition.job_names, batch_definition.all_jobs)
```

Comment:
The reprocess_recipe method calls the queue method internally. Using this method instead allows us to reuse the logic for populating job names, building the recipe graph deltas, and the unit test coverage for making sure all the right jobs get created.